experimental/adaptive-card-prompt-sample/dotnet/AdaptiveCardPrompt.cs

```diff
@@ -0,0 +1,294 @@
+            }
+        }
+
+        private Attachment AddPromptIdToCard(Attachment card)
+        {
+            card.Content = DeepSearchJsonForActionsAndAddPromptId((JObject)card.Content);
+            return card;
+        }
+
+        private JObject DeepSearchJsonForActionsAndAddPromptId(JObject json)
+        {
+            var submitAction = "Action.Submit";
+            var showCardAction = "Action.ShowCard";
+
+            foreach (var obj in json)
+            {
+                // Search for all submits in actions
+                if (obj.Key == "actions")
+                {
+                    // Must convert json to list so that we can make changes to it
+                    var i = 0;
+                    foreach (var action in (obj.Value as JArray).ToObject<List<JObject>>())
+                    {
+                        if (!string.IsNullOrWhiteSpace(action["type"].ToString()) && action["type"].ToString() == submitAction)
+                        {
+                            // Make changes to original json, not the created list
+                            json[obj.Key][i]["data"] = json[obj.Key][i]["data"] ?? new JObject();
+                            json[obj.Key][i]["data"]["promptId"] = _promptId;
+
+                            // Recursively search Action.ShowCard for Submits within the nested card
+                        }
+                        else if (!string.IsNullOrWhiteSpace(action["type"].ToString()) && action["type"].ToString() == showCardAction)
+                        {
+                            json[obj.Key][i] = DeepSearchJsonForActionsAndAddPromptId(action);
+                        }
+
+                        i++;
+                    }
+                }
+
+                // Search for all submits in selectActions
+                else if (obj.Key == "selectAction")
+                {
+                    if (!string.IsNullOrWhiteSpace(obj.Value["type"].ToString()) && obj.Value["type"].ToString() == submitAction)
+                    {
+                        json[obj.Key]["data"] = obj.Value["data"] ?? new JObject();
+                        json[obj.Key]["data"]["promptId"] = _promptId;
+                    }
+
+                    // Recursively search Action.ShowCard for Submits within the nested card
+                    if (!string.IsNullOrWhiteSpace(obj.Value["type"].ToString()) && obj.Value["type"].ToString() == showCardAction)
+                    {
+                        json[obj.Key] = DeepSearchJsonForActionsAndAddPromptId(obj.Value as JObject);
+                    }
+                }
+
+                // Recursively search all other objects
+                else if (!string.IsNullOrWhiteSpace(obj.Value.ToString()) && obj.Value is JObject)
+                {
+                    json[obj.Key] = DeepSearchJsonForActionsAndAddPromptId(obj.Value as JObject);
+                }
+            }
+
+            return json;
+        }
```

Comment:
@stevengum 

I agree that it's nasty. Even using `AdaptiveCard.FromJson()`, I think this is necessary.

You probably gathered this part, but basically, in an Adaptive Card's JSON, you send data by having a `data` object within `Action.Submit`. If we add the feature that ensures that user input comes from the correct card, we need to add the cardId to the `data` object so that we know where it's coming from.

The problem that this code solves is that `Action.Submit` can be nested shallow or deep in a LOT of different locations. I don't know of a better way to handle adding a recursive deep-searching JSON algorithm. [It's a *little* cleaner in the new Sample PR](https://github.com/microsoft/BotBuilder-Samples/pull/1986/files#diff-552617ffac23815c464228fb53b84f05R187), but still kinda nasty.

Really, really open to hearing your thoughts on this.

Reply:
I think we as a team need to decide on what functionality this prompt addresses ([related comment](https://github.com/microsoft/BotBuilder-Samples/pull/1616#discussion_r347053880)). 

After we do that, we move into territory of implementation. Does the developer set the Adaptive Card in the `constructor()`, or can the they pass into the prompt an Activity with an Adaptive Card?

If the card is only set in the constructor, then the card is specifically tied to the prompt, so the developer knows the prompt ID ahead of time and can pass it to the Action.Submit. 

Traversing the card sounds awful in any respect, so I'd prefer to avoid that responsibility if possible, and instead cleanly document what the user needs to do.

That said, perhaps we can still manage the values from the library:

#### [AdaptiveCards Library](https://github.com/microsoft/AdaptiveCards/tree/master/source/dotnet/Library/AdaptiveCards)
  - [AdaptiveSubmitAction](https://github.com/microsoft/AdaptiveCards/blob/master/source/dotnet/Library/AdaptiveCards/AdaptiveSubmitAction.cs)

Reply:
> If the card is only set in the constructor, then the card is specifically tied to the prompt, so the developer knows the prompt ID ahead of time and can pass it to the Action.Submit.

> Traversing the card sounds awful in any respect, so I'd prefer to avoid that responsibility if possible, and instead cleanly document what the user needs to do.

Yeah. I'll take all of that out and just make them put it in.

I think, this, and a lot of the issues you're finding with this are because this was developed first from a standpoint of "this would be fun; let's see what I can do and then give it to the SDK" vs. "let's write something good for the SDK". Originally, I really had no intention of submitting it, but then we kept getting questions related to it.

---

That AdaptiveSubmitAction is interesting. From the description, it looks like it could do this for us, but [the only examples I can find of it](https://github.com/microsoft/AdaptiveCards/blob/cb28506ed9fc4b201badcefa4f7811fc340294c7/source/uwp/UWPUnitTests/ObjectModelTests.cs) show that the user/dev would need to use it, otherwise we're back to deep searching.

I think that because:

1. It would add another dependency
2. We can solve this via your suggestion above

It's probably not necessary.